Update filesystem.py

lchantre opened the pull request:
Fixes #35509 
add ocfs2 support for module filesystem

 - Feature Pull Request

module filesystem

ansible devel

lchantre committed:
Update filesystem.py

#35509
add ocfs2 support for module filesystem

lchantre committed:
Update filesystem.py

add ocfs2 support for module filesystem
Conflicts:
	lib/ansible/modules/system/filesystem.py

lchantre committed:
Update filesystem.py

add ocfs2 support for module filesystem
Conflicts:
	lib/ansible/modules/system/filesystem.py

pilou- reviewed (commented):
Could you include integration tests for this filesystem ?
Integration tests for this module are located there: https://github.com/ansible/ansible/tree/devel/test/integration/targets/filesystem.
Don't hesitate to ask if you need some pointers/doc.

lchantre committed:
Update setup.yml

#35509 add ocfs2 support for module filesystem

pilou- commented:
Could you append the commit adding the test for ocfs2 to the branch of this PR instead of creating another PR ? This way, the CI would be able to check this new feature before the merge :)

lchantre commented:
Hello
I'm sorry I don't know how to add test to the branch of PR #35551

pilou- commented:
The following commands should work:
```
cd ansible # work from your local copy
git checkout patch-1 # work on the first branch
git cherry-pick 0720536b04df5256278e304bdc86af02e9792111 # this will 'copy' the commit to the current branch
git cherry-pick 056e12cbced1d3ed8a5741adb608ce178fdef0d2 # the same for the third commit
git push # update your remote branch
```

If needed, new commits can be added using:
```
editor lib/ansible/modules/system/filesystem.py # make some updates
git add lib/ansible/modules/system/filesystem.py  # take in account the updates
git commit # create a commit on the local branch
git push # update the remote branch
```

lchantre committed:
Update filesystem.py
Conflicts:
	lib/ansible/modules/system/filesystem.py

lchantre committed:
delete trailing spaces

Conflicts:
	lib/ansible/modules/system/filesystem.py

lchantre commented:
Hello,
many thanks for you help
I corrected whitespace and apply following commands :
cd ansible # work from your local copy
git checkout patch-1 # work on the first branch
git cherry-pick 0720536b04df5256278e304bdc86af02e9792111 # this will 'copy' the commit to the current branch
git cherry-pick 056e12cbced1d3ed8a5741adb608ce178fdef0d2 # the same for the third commit
git push # update your remote branch
and it seems to be OK on my side

Let me know if it's ok from you side

regards

lchantre committed:
add ocsfs2 defaut var

Conflicts:
	test/integration/targets/filesystem/defaults/main.yml

mattclay commented:
CI failure in integration tests: 

`failure: Failed to find required executable mkfs.ocfs2 in paths: /tmp/ansible-test-coverage-Yiby6_/coverage:/root/ansible/bin:/usr/local/sbin:/sbin:/bin:/usr/sbin:/usr/bin:/root/bin`

lchantre committed:
Install ocfs2-tools for all linux

lchantre commented:
Wand to install ocfs2.tools only for Debian distribution now change to install it for all linux in the test

pilou- commented:
* CentOS/RHEL 6 and 7: ocfs2-tools isn't available: [ref1](https://access.redhat.com/solutions/38913), [ref2](https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/7/html/package_manifest/chap-base-server-variant) (there is one [alternative](http://www.shapeblue.com/installing-and-configuring-an-ocfs2-clustered-file-system/))
* FreeBSD: ocfs2-tools doesn't seem available
* Ubuntu (all versions): ocfs2 filesystem isn't recognized by blkid utility:

  ```
  TASK [filesystem : filesystem creation] ***
  changed: [testhost] => {"changed": true, "failed": false}
  
  TASK [filesystem : assert] ***
  ok: [testhost] => {
      "changed": false, 
      "failed": false, 
      "msg": "All assertions passed"
  }
  
  TASK [filesystem : command] ***
  fatal: [testhost]: FAILED! => {"changed": true, "cmd": ["blkid", "-c", "/dev/null", "-o", "value", "-s", "UUID", "/root/ansible_testing/img"], "delta": "0:00:00.032593", "end": "2018-02-05 08:22:50.137090", "failed": true, "msg": "non-zero return code", "rc": 2, "start": "2018-02-05 08:22:50.104497", "stderr": "", "stderr_lines": [], "stdout": "", "stdout_lines": []}
  ```
* OpenSuse 42.3
  ```
  TASK [filesystem : filesystem creation] ***
  fatal: [testhost]: FAILED! => {"changed": false, "cmd": "/sbin/mkfs.ocfs2 -f /root/ansible_testing/img", "failed": true, "msg": "/sbin/mkfs.ocfs2: invalid option -- 'f'\nusage: mkfs.ocfs2 [-b block-size] [-C cluster-size] [-J journal-options]\n\t\t[-L volume-label] [-M mount-type] [-N number-of-node-slots]\n\t\t[-T filesystem-type] [-U uuid][-HFnqvV] [--dry-run]\n\t\t[--fs-feature-level=[default|max-compat|max-features]] \n\t\t[--fs-features=[[no]sparse,...]] [--global-heartbeat]\n\t\t[--cluster-stack=stackname] [--cluster-name=clustername]\n\t\t[--no-backup-super] device [blocks-count]", "rc": 1, "stderr": "/sbin/mkfs.ocfs2: invalid option -- 'f'\nusage: mkfs.ocfs2 [-b block-size] [-C cluster-size] [-J journal-options]\n\t\t[-L volume-label] [-M mount-type] [-N number-of-node-slots]\n\t\t[-T filesystem-type] [-U uuid][-HFnqvV] [--dry-run]\n\t\t[--fs-feature-level=[default|max-compat|max-features]] \n\t\t[--fs-features=[[no]sparse,...]] [--global-heartbeat]\n\t\t[--cluster-stack=stackname] [--cluster-name=clustername]\n\t\t[--no-backup-super] device [blocks-count]\n", "stderr_lines": ["/sbin/mkfs.ocfs2: invalid option -- 'f'", "usage: mkfs.ocfs2 [-b block-size] [-C cluster-size] [-J journal-options]", "\t\t[-L volume-label] [-M mount-type] [-N number-of-node-slots]", "\t\t[-T filesystem-type] [-U uuid][-HFnqvV] [--dry-run]", "\t\t[--fs-feature-level=[default|max-compat|max-features]] ", "\t\t[--fs-features=[[no]sparse,...]] [--global-heartbeat]", "\t\t[--cluster-stack=stackname] [--cluster-name=clustername]", "\t\t[--no-backup-super] device [blocks-count]"], "stdout": "", "stdout_lines": []}
  ```

@lchantre 
1. Tests could be enabled for Ubuntu only, it would be sufficient.
1. `blkid` isn't able to handle `ocfs2` filesystems, other updates to `filesystem` module are required.

lchantre committed:
Testing ocfs2 on for Ubuntu - restrict blkid to be be done

lchantre committed:
Update filesystem.py

delete trailing spaces
Conflicts:
	lib/ansible/modules/system/filesystem.py

lchantre commented:
Hello,
Sorry I couldn't find anymore how to git clone my own version of ansible to rebase my branch :-((
I only find the clone link https://github.com/ansible/ansible.git, could you help me ?
best regards

lchantre commented:
Hello,
I find it ! git clone https://github.com/lchantre/ansible.git   !
I will try to rebase it

lchantre commented:
I follow the instructions but I hava a problem at the end
luc@sv1:~/git/ansible$ git pull --rebase upstream devel
From https://github.com/ansible/ansible
 * branch            devel      -> FETCH_HEAD
Current branch HEAD is up to date.

luc@sv1:~/git/ansible$ git status
HEAD detached from refs/heads/patch-1
nothing to commit, working directory clean

uc@sv1:~/git/ansible$ git push --force
fatal: You are not currently on a branch.
To push the history leading to the current (detached HEAD)
state now, use

    git push origin HEAD:<name-of-remote-branch>

pilou- commented:
@lchantre could you try these commands:

```sh
git checkout patch-1
git branch -m patch-1.old # rename the old branch
git checkout devel
git pull upstream
git checkout -b patch-1 # create a new one
git cherry-pick 0720536b04df5256278e304bdc86af02e9792111^..abd2d03ea86ab708436528335311f7cb2b5e7198
git cherry-pick 02a7eb4457323b27ccabb542ba4dc11f0d0f7451 # apply existing commits (but not the merge commit)
git push -f -u origin patch-1 # update remote branch
```

lchantre commented:
Hello,
some problems with the following command

git pull upstream
fatal: 'upstream' does not appear to be a git repository
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.

pilou- commented:
@lchantre could you paste the output of the following command `git remote -v` ?

lchantre commented:
origin	https://github.com/lchantre/ansible.git (fetch)
origin	https://github.com/lchantre/ansible.git (push)

pilou- commented:
Just use this command before the ones mentioned in the [previous comment](https://github.com/ansible/ansible/pull/35551#issuecomment-374807400):

```sh
git remote add upstream https://github.com/ansible/ansible.git 
```

There is some related documentation in the [developer guide](http://docs.ansible.com/ansible/2.5/dev_guide/developing_rebasing.html).

lchantre commented:
Hello
after git remote add upstream https://github.com/ansible/ansible.git 

git remote -v 
origin	https://github.com/lchantre/ansible.git (fetch)
origin	https://github.com/lchantre/ansible.git (push)
upstream	https://github.com/ansible/ansible.git (fetch)
upstream	https://github.com/ansible/ansible.git (push)

I tried to pull upstream (git pull upstream) and I had this message
remote: Counting objects: 13091, done.
remote: Compressing objects: 100% (14/14), done.
remote: Total 13091 (delta 4253), reused 4260 (delta 4251), pack-reused 8826
Receiving objects: 100% (13091/13091), 8.05 MiB | 866.00 KiB/s, done.
Resolving deltas: 100% (8564/8564), completed with 2020 local objects.
From https://github.com/ansible/ansible
 * [new branch]      devel      -> upstream/devel
...
You asked to pull from the remote 'upstream', but did not specify
a branch. Because this is not the default configured remote
for your current branch, you must specify a branch on the command line.

I try to checkout and cherry-pick ?

pilou- commented:
I updated one command: use `git checkout upstream/devel -b devel` instead of `git checkout devel`.

lchantre commented:
I'm sorry but :
:~/ansible$ git checkout upstream/devel -b devel
fatal: Cannot update paths and switch to branch 'devel' at the same time.
Did you intend to checkout 'upstream/devel' which can not be resolved as commit?

pilou- commented:
Could you try to use `git fetch upstream` before this checkout ?

lchantre commented:
hello
sorry again
git checkout upstream/devel -b devel
fatal: A branch named 'devel' already exists.

pilou- commented:
This command should work: `git checkout upstream/devel -b patch-1`; after that use the cherry-pick command.

lchantre commented:
ok many thanks

lchantre commented:
I hope it's ok now

pilou- commented:
You need to remove these lines:
```
+>>>>>>>
+========
+<<<<<<< HEAD
```
from
- lib/ansible/modules/system/filesystem.py
- test/integration/targets/filesystem/defaults/main.yml

Commit:
remove bad lines

Commit:
add new line and update choice documentation

pilou- commented on lib/ansible/modules/system/filesystem.py:
```diff
@@ -22,7 +22,7 @@
-    choices: [ btrfs, ext2, ext3, ext4, ext4dev, lvm, reiserfs, xfs, vfat ]
+    choices: [ btrfs, ext2, ext3, ext4, ext4dev, lvm, reiserfs, xfs, vfat, ocfs2 ]
     description:
     - Filesystem type to be created.
```
@lchantre could you add a line in `description` mentioning something like `ocfs2 support was added in 2.6`. ?

Commit:
adding description for ocfs2 support

pilou- commented on test/integration/targets/filesystem/tasks/main.yml:
```diff
@@ -12,6 +12,7 @@
     grow: '{{ item.value.grow }}'
   when:
     - 'not (item.key == "btrfs" and ansible_system == "FreeBSD")'
+    - 'not (item.key == "ocfs2" and ansible_system != "Ubuntu")'
```
Tests are currently not run by the CI because this condition is always `False`, `ansible_distribution` must be used instead of `ansible_system`. Note that `not (item.key == "ocfs2" and ansible_os_family != 'Debian')` could be used too.

pilou- commented on test/integration/targets/filesystem/tasks/setup.yml:
```diff
@@ -29,6 +29,12 @@
+    package:
+      name: ocfs2-tools
+      state: present
+    when: ansible_os_family == 'Ubuntu'
```
`ansible_distribution` must be used instead of `ansible_os_family`. Note that not `ansible_os_family == 'Debian'` could be used too.

pilou- reviewed (commented):
Thank you for your responsiveness ! Tests are not run in the CI, this must be fixed in order to be able to merge.

Commit:
use correct variable ansible_distribution to test ocfs2 with Debian
distribution

Commit:
use ansible_os_family for Debian

pilou- commented:
Filesystem creation fails:
```
2018-04-10 08:33:12     "stderr_lines": [
2018-04-10 08:33:12         "mkfs.ocfs2 1.6.4", 
2018-04-10 08:33:12         "mkfs.ocfs2: Unable to find available bit while creating lost+found"
2018-04-10 08:33:12     ], 

2018-04-10 08:33:12     "stdout_lines": [
2018-04-10 08:33:12         "Cluster stack: classic o2cb", 
2018-04-10 08:33:12         "Label: ", 
2018-04-10 08:33:12         "Features: sparse backup-super unwritten inline-data strict-journal-super xattr", 
2018-04-10 08:33:12         "Block size: 1024 (10 bits)", 
2018-04-10 08:33:12         "Cluster size: 4096 (12 bits)", 
2018-04-10 08:33:12         "Volume size: 10485760 (2560 clusters) (10240 blocks)", 
2018-04-10 08:33:12         "Cluster groups: 1 (tail covers 2560 clusters, rest cover 2560 clusters)", 
2018-04-10 08:33:12         "Extent allocator size: 0 (0 groups)", 
2018-04-10 08:33:12         "Journal size: 4194304", 
2018-04-10 08:33:12         "Node slots: 2", 
2018-04-10 08:33:12         "Creating bitmaps: done", 
2018-04-10 08:33:12         "Initializing superblock: done", 
2018-04-10 08:33:12         "Writing system files: done", 
2018-04-10 08:33:12         "Writing superblock: done", 
2018-04-10 08:33:12         "Writing backup superblock: 0 block(s)", 
2018-04-10 08:33:12         "Formatting Journals: done", 
2018-04-10 08:33:12         "Growing extent allocator: done", 
2018-04-10 08:33:12         "Formatting slot map: done", 
2018-04-10 08:33:12         "Formatting quota files: done", 
2018-04-10 08:33:12         "Writing lost+found: "
2018-04-10 08:33:12     ]
```

@lchantre i am able to reproduce this error, it seems the minimal size is 11Mo, not [10Mo](https://github.com/ansible/ansible/pull/35551/files#diff-cd0480973fb28a7c2d6e1c5a237df74eR15).

Commit:
increase ocfs2 fs size to 20M (minimal size 11 instead of 10M)

pilou- commented:
@lchantre 
- on Ubuntu 14.04, test fails because `blkid` doesn't recognize `ocfs2` filesystem. Filesystem module requires that `blkid` is able to identify created filesystem.
- on Ubuntu 16.04, test fails because `Ocfs2.MKFS_FORCE_FLAGS` isn't set. But `mkfs.ocfs2 -F` won't work because mkfs.ocfs2 asks for a confirmation:
  ```
  $ mkfs.ocfs2 -F img 
  mkfs.ocfs2 1.6.4
  Cluster stack: classic o2cb
  Overwriting existing ocfs2 partition.
  WARNING: Cluster check disabled.
  Proceed (y/N):
  ```

  The undocumented `x` switch must be used too (`Ocfs2.MKFS_FORCE_FLAGS` should equal to `-Fx`).

pilou- committed:
filesystem tests: update ocfs2 fs size on Trusty

Using Ubuntu 14.04, test fails because 'blkid' < 2.21 doesn't recognize
'ocfs2' filesystem smaller than 108Mo:
https://github.com/karelzak/util-linux/commit/6baa15039873aac5513abde621f60c63537b88e9

pilou- committed:
filesystem: fix MKFS_FORCE_FLAGS for ocfs2

mkfs.ocfs2 -F won't work because mkfs.ocfs2 asks for a confirmation:

    $ mkfs.ocfs2 -F img
    mkfs.ocfs2 1.6.4
    Cluster stack: classic o2cb
    Overwriting existing ocfs2 partition.
    WARNING: Cluster check disabled.
    Proceed (y/N):

The undocumented 'x' switch must be used too.

pilou- commented:
LGTM

pilou- commented:
@Akasurde please, could you take a look at this one ? One approval is still required.
